rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+            CStr::from_ptr(message).to_str().unwrap().to_string()
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Copy, Clone, Debug)]
+pub struct Comment(*mut yp_comment_t);
+
+impl Comment {
+    /// Returns the text of the comment.
+    #[must_use]
+    pub fn text(&self) -> String {
+        unsafe {
+            let start: *mut u8 = (*(self.0)).start as *mut u8;
+            let end: *mut u8 = (*(self.0)).end as *mut u8;
+
+            let length = end as usize - start as usize;
+            String::from_raw_parts(start, length, length)
+        }
+    }
+}
+
+/// A couple of different concepts are represented as linked lists in the
+/// parser. This includes errors, warnings, and comments. We want to be able to
+/// iterate over them in a consistent fashion. As such, we define a trait that
+/// can be used to iterate over them.
+pub trait ListNode<T> {
+    /// Returns true if the node is null.
+    fn is_null(&self) -> bool;
+
+    /// Returns the next node in the list.
+    fn next(&self) -> T;
+}
+
+impl ListNode<Self> for Diagnostic {
+    fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    fn next(&self) -> Self {
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_diagnostic_t>() })
+    }
+}
+
+impl ListNode<Self> for Comment {
+    fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    fn next(&self) -> Self {
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_comment_t>() })
+    }
+}
+
+/// An iterator that can be used to iterate over a linked list.
+pub struct ListIter<T: ListNode<T> + Copy> {
+    current: T
+}
+
+impl<T: ListNode<T> + Copy> Iterator for ListIter<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            None
+        } else {
+            let current = self.current;
+            self.current = self.current.next();
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult {
+    source: CString,
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl ParseResult {
+    /// Returns the source string that was parsed.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the source string is not able to be converted into a `&str`.
+    #[must_use]
+    pub fn source(&self) -> String {
+        self.source.to_str().unwrap().to_string()
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors_iter(&self) -> ListIter<Diagnostic> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            ListIter { current: Diagnostic(list.head.cast::<yp_diagnostic_t>()) }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings_iter(&self) -> ListIter<Diagnostic> {
```

Comment:
generally, methods like these would be called `errors` or `warnings` and return iterator types called `Errors` and `Warnings`.

See `str::bytes` or `[_]::chunks` in `std` for examples.

Reply:
these iterators should also probably be bound to the lifetime of the parse result